driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -0,0 +1,313 @@
+import com.mongodb.internal.operation.AsyncOperationHelper.AsyncCallableConnectionWithCallback;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonDocument;
+import org.bson.BsonTimestamp;
+import org.bson.BsonValue;
+import org.bson.codecs.BsonDocumentCodec;
+import org.bson.codecs.Decoder;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.mongodb.assertions.Assertions.assertNotNull;
+import static com.mongodb.assertions.Assertions.assertTrue;
+import static com.mongodb.assertions.Assertions.doesNotThrow;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.FIRST_BATCH;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.MESSAGE_IF_CLOSED_AS_CURSOR;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.NEXT_BATCH;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.NO_OP_FIELD_NAME_VALIDATOR;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.getKillCursorsCommand;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.getMoreCommandDocument;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.logCommandCursorResult;
+import static com.mongodb.internal.operation.CommandBatchCursorHelper.translateCommandException;
+import static java.util.Collections.emptyList;
+
+class AsyncCommandBatchCursor<T> implements AsyncAggregateResponseBatchCursor<T> {
+
+    private final MongoNamespace namespace;
+    private final long maxTimeMS;
+    private final Decoder<T> decoder;
+    @Nullable
+    private final BsonValue comment;
+    private final int maxWireVersion;
+    private final boolean firstBatchEmpty;
+    private final ResourceManager resourceManager;
+    private final AtomicBoolean processedInitial = new AtomicBoolean();
+    private int batchSize;
+    private volatile CommandCursorResult<T> commandCursorResult;
+
+    AsyncCommandBatchCursor(
+            final BsonDocument commandCursorDocument,
+            final int batchSize, final long maxTimeMS,
+            final Decoder<T> decoder,
+            @Nullable final BsonValue comment,
+            final AsyncConnectionSource connectionSource,
+            final AsyncConnection connection) {
+        ConnectionDescription connectionDescription = connection.getDescription();
+        this.commandCursorResult = toCommandCursorResult(connectionDescription.getServerAddress(), FIRST_BATCH, commandCursorDocument);
+        this.namespace = commandCursorResult.getNamespace();
+        this.batchSize = batchSize;
+        this.maxTimeMS = maxTimeMS;
+        this.decoder = decoder;
+        this.comment = comment;
+        this.maxWireVersion = connectionDescription.getMaxWireVersion();
+        this.firstBatchEmpty = commandCursorResult.getResults().isEmpty();
+
+        AsyncConnection connectionToPin = connectionSource.getServerDescription().getType() == ServerType.LOAD_BALANCER
```

Comment:
I think this is equally good check for LB, but I thought the old way is fine too.  What was sketchy about it?

Reply:
Turns out `connectionDescription.getServerType()` would consistently fail and not be equal to `ServerType.LOAD_BALANCER`. 

So this was the fix.

Reply:
The build before this fix: https://spruce.mongodb.com/version/654a27dd0305b9e381425fdf/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

Reply:
I thought the fact that they reported different ServerTypes was sketchy.

Reply:
Oh, so this reverts back to the way the check is on master.  Got it.